File: chapter_02_排序/4.merge_sort.go
package main

import (
	"fmt"
	"math"
)

func merge(data IntSlice, lo int, mid int, hi int) {
	var (
		i, j = lo, mid+1
		aux = IntSlice(make([]int, hi+1-lo))
	)
	copy(aux, data[lo:hi+1])

	// 将元素复制一份出来
	for k:=lo; k<=hi; k++ {
		if i > mid { // 左边用完了, 取右边
			data[k] = aux[j-lo]
			j++
		} else if j > hi { // 右边用完了，取左边的
			data[k] = aux[i-lo]
			i++
		} else if aux.Less(i-lo, j-lo) { // 取 i
			data[k] = aux[i-lo]
			i++
		} else {
			data[k] = aux[j-lo]
			j++
		}
	}
	fmt.Println(lo, mid, hi, aux, data)
}

// 将元素氛围两部分排序 [lo, mid] (mid, hi]
func mergeSort(data IntSlice, lo int, hi int) {
	if hi <= lo {
		return
	}
	mid := (hi+lo) / 2
	mergeSort(data, lo, mid)
	mergeSort(data, mid+1, hi)
	merge(data, lo, mid, hi)
}

func MergeSort(data IntSlice) {
	mergeSort(data, 0, data.Len()-1)
}

/*
   自第向下 排序
 */
func MergeSortFromBottomToTop(data IntSlice) {
	l := data.Len()
	for sz:=1; sz<l; sz+=sz { // 1, 2, 4, 8
		for i:=0; i<l-sz; i+=2*sz {
			merge(data, i, i+sz-1, int(math.Min(float64(i+2*sz-1), float64(l-1))))
		}
	}
}

func main() {
	p := generateIntSlice(999, 10)
	//fmt.Println("归并排序前", p)
	//MergeSort(p)
	//fmt.Println("归并排序后：", p, p.IsSorted())

	p.Shuffle()
	MergeSortFromBottomToTop(p)
	fmt.Println(p, p.IsSorted())
}

File: chapter_02_排序/py/4_merge_sort.py
# coding: utf-8

"""
归并排序. 自顶向下topDown，自底向上downTop
"""
import sys
import time
from sort_base import DataInterface, IntData, generateIntData, time_it


class MergeSort(object):
    """递归排序
    然后将字数组的排序结果，合并为有序的数组
    """

    def __init__(self, data):
        self.data = data

    def _merge(self, lo, mid, hi):
        # 1/ 备份
        aux = []
        for i in xrange(lo, hi+1):
            aux.append(self.data.data[i])  # 直接访问内部变量，不提倡

        # 2/ 合并
        # [i, mid] 有序 [mid+1, hi] 有序 现在将他们合并
        i, j = lo, mid+1
        for k in xrange(lo, hi+1):
            if i > mid: # 左边用完了,
                self.data.data[k] = aux[j-lo]
                j += 1
            elif j > hi: # 右边用完了
                self.data.data[k] = aux[i-lo]
                i += 1
            elif aux[i-lo] > aux[j-lo]: # 取右边的
                self.data.data[k] = aux[j-lo]
                j += 1
            else:
                self.data.data[k] = aux[i - lo]
                i += 1

    def _top_down(self, lo, hi):
        if lo >= hi:
            return
        mid = (lo+hi)/2
        self._top_down(lo, mid)
        self._top_down(mid+1, hi)
        self._merge(lo, mid, hi)

    @staticmethod
    @time_it
    def top_down(data):
        """自顶向下"""
        o = MergeSort(data)
        o._top_down(0, data.len()-1)

    @staticmethod
    @time_it
    def down_top(data):
        """自底向下
        最开始排定每2个，之后4个，之后8个
        """
        o = MergeSort(data)
        sz = 1
        l = data.len()
        while sz < l:
            for i in range(0, l-sz+1, sz*2):
                o._merge(i, i+sz-1, min(i+2*sz-1, l-1))
            sz *= 2


if __name__ == '__main__':
    sys.setrecursionlimit(1000000)
    d = generateIntData(100000, 100)

    print d.is_sorted()

    MergeSort.top_down(d)
    print d.is_sorted()

    d.shuffle()
    MergeSort.down_top(d)
    print d.is_sorted()

